From 171ea60b1d897100dbbea1015b294905e9039205 Mon Sep 17 00:00:00 2001
From: Marko Katic <dromede.gmail.com>
Date: Tue, 21 Feb 2012 22:25:41 +0100
Subject: [PATCH 14/15] invert usb host gpio.

---
 arch/arm/mach-pxa/spitz.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-pxa/spitz.c b/arch/arm/mach-pxa/spitz.c
index 194f852..37b1a37 100644
--- a/arch/arm/mach-pxa/spitz.c
+++ b/arch/arm/mach-pxa/spitz.c
@@ -439,6 +439,7 @@ static struct gpio_keys_button spitz_gpio_keys[] = {
 	{
 		.type	= EV_SW,
 		.code	= 3,
+		.active_low = 1,
 		.gpio	= SPITZ_GPIO_USB_CONNECT,
 		.desc	= "USB host cable connected",
 	},
-- 
1.7.4.1

