Index: pxaregs-1.14-r0/pxaregs.c
===================================================================
--- pxaregs-1.14-r0.orig/pxaregs.c	2012-02-14 19:31:07.000000000 +0100
+++ pxaregs-1.14-r0/pxaregs.c	2012-02-14 19:31:18.000000000 +0100
@@ -293,6 +293,14 @@
 { "CKEN_13",   0x41300004, 13, 0x00000001, 'd', "CM FIPC clock enabled" },
 { "CKEN_14",   0x41300004, 14, 0x00000001, 'd', "CM I2C clock enabled" },
 { "CKEN_16",   0x41300004, 16, 0x00000001, 'd', "CM LCD clock enabled" },
+{ "CKEN_17",   0x41300004, 17, 0x00000001, 'd', "MSL clock enable" }, 
+{ "CKEN_18",   0x41300004, 18, 0x00000001, 'd', "USIM clock enable" }, 
+{ "CKEN_19",   0x41300004, 19, 0x00000001, 'd', "Keypad clock enable" }, 
+{ "CKEN_20",   0x41300004, 20, 0x00000001, 'd', "Internal Memory clock enable" }, 
+{ "CKEN_21",   0x41300004, 21, 0x00000001, 'd', "Memeory Stick clock enable" }, 
+{ "CKEN_22",   0x41300004, 22, 0x00000001, 'd', "Memory Controller clock enable" }, 
+{ "CKEN_23",   0x41300004, 23, 0x00000001, 'd', "SSP1 clock enable" }, 
+{ "CKEN_24",   0x41300004, 24, 0x00000001, 'd', "Quick Capture clock enable" }, 
 
 { "OSCC",      0x41300008,  0, 0xffffffff, 'x', "Oscillator Configuration Register (3-38)" },
 { "OSCC_OOK",  0x41300008,  0, 0x00000001, 'd', "CM 32.768 kHz oscillator enabled and stabilized" },
