Index: git/src/gtk/fm-places-model.c
===================================================================
--- git.orig/src/gtk/fm-places-model.c	2012-02-10 00:36:39.000000000 +0100
+++ git/src/gtk/fm-places-model.c	2012-02-10 00:37:45.000000000 +0100
@@ -493,7 +493,8 @@
     gtk_list_store_append(model, &it);
     pix = fm_icon_get_pixbuf(item->fi->icon, fm_config->pane_icon_size);
     gtk_list_store_set(model, &it, FM_PLACES_MODEL_COL_ICON, pix, FM_PLACES_MODEL_COL_LABEL, item->fi->path->name, FM_PLACES_MODEL_COL_INFO, item, -1);
-    g_object_unref(pix);
+    if (pix)
+		g_object_unref(pix);
     fm_file_info_job_add(job, item->fi->path);
 
     /* Only show desktop in side pane when the user has a desktop dir. */
@@ -507,7 +508,8 @@
         gtk_list_store_append(model, &it);
         pix = fm_icon_get_pixbuf(item->fi->icon, fm_config->pane_icon_size);
         gtk_list_store_set(model, &it, FM_PLACES_MODEL_COL_ICON, pix, FM_PLACES_MODEL_COL_LABEL, _("Desktop"), FM_PLACES_MODEL_COL_INFO, item, -1);
-        g_object_unref(pix);
+        if (pix)
+			g_object_unref(pix);
         fm_file_info_job_add(job, item->fi->path);
     }
 
@@ -522,7 +524,8 @@
     gtk_list_store_append(model, &it);
     pix = fm_icon_get_pixbuf(item->fi->icon, fm_config->pane_icon_size);
     gtk_list_store_set(model, &it, FM_PLACES_MODEL_COL_ICON, pix, FM_PLACES_MODEL_COL_LABEL, _("Applications"), FM_PLACES_MODEL_COL_INFO, item, -1);
-    g_object_unref(pix);
+    if (pix)
+		g_object_unref(pix);
     /* fm_file_info_job_add(job, item->fi->path); */
 
     /* volumes */
