Index: git/data/ui/Makefile.am
===================================================================
--- git.orig/data/ui/Makefile.am	2012-01-12 15:48:57.000000000 +0100
+++ git/data/ui/Makefile.am	2012-01-12 15:49:24.000000000 +0100
@@ -20,4 +20,4 @@
 # Purge GtkBuilder UI files
 %.ui: %.glade
 	cp $< $@
-	$(top_builddir)/src/xml-purge $@
+	
Index: git/src/Makefile.am
===================================================================
--- git.orig/src/Makefile.am	2012-01-12 15:49:01.000000000 +0100
+++ git/src/Makefile.am	2012-01-12 15:49:12.000000000 +0100
@@ -401,8 +401,3 @@
 #        -no-undefined \
 #        -export-symbols-regex gtk
 
-# Little program to optimize size of xml files
-noinst_PROGRAMS=xml-purge
-xml_purge_SOURCES=xml-purge.c
-xml_purge_CFLAGS=$(GIO_CFLAGS)
-xml_purge_LDADD=$(GIO_LIBS)
