Index: tslib-1.0/plugins/linear.c
===================================================================
--- tslib-1.0.orig/plugins/linear.c	2012-02-12 01:03:12.000000000 +0100
+++ tslib-1.0/plugins/linear.c	2012-02-12 01:04:17.000000000 +0100
@@ -57,9 +57,10 @@
 			samp->y =	( lin->a[5] +
 					lin->a[3]*xtemp +
 					lin->a[4]*ytemp ) / lin->a[6];
-
-			samp->pressure = ((samp->pressure + lin->p_offset)
-						 * lin->p_mult) / lin->p_div;
+			
+			if ((samp->x < 0) || (samp->y < 0))
+				return 0;
+			
 			if (lin->swap_xy) {
 				int tmp = samp->x;
 				samp->x = samp->y;
