diff --git a/data/gui/preferences.ui b/data/gui/preferences.ui
index 9c6588d..bda0909 100644
--- a/data/gui/preferences.ui
+++ b/data/gui/preferences.ui
@@ -8,18 +8,6 @@
       <column type="gchararray"/>
     </columns>
   </object>
-  <object class="GtkListStore" id="hotkey_binding_model">
-    <columns>
-      <!-- column-name hotkey_name_column -->
-      <column type="gchararray"/>
-      <!-- column-name hotkey_value_column -->
-      <column type="gchararray"/>
-      <!-- column-name hotkey_enum_column -->
-      <column type="gint"/>
-      <!-- column-name hotkey_enabled_column -->
-      <column type="gboolean"/>
-    </columns>
-  </object>
   <object class="GtkListStore" id="theme_name_model">
     <columns>
       <!-- column-name theme_name_column -->
@@ -204,6 +192,8 @@
                                         <property name="invisible_char">●</property>
                                         <property name="primary_icon_activatable">False</property>
                                         <property name="secondary_icon_activatable">False</property>
+                                        <property name="primary_icon_sensitive">True</property>
+                                        <property name="secondary_icon_sensitive">True</property>
                                         <property name="adjustment">volume_adjustment</property>
                                         <property name="numeric">True</property>
                                       </object>
@@ -219,6 +209,8 @@
                                         <property name="invisible_char">●</property>
                                         <property name="primary_icon_activatable">False</property>
                                         <property name="secondary_icon_activatable">False</property>
+                                        <property name="primary_icon_sensitive">True</property>
+                                        <property name="secondary_icon_sensitive">True</property>
                                       </object>
                                       <packing>
                                         <property name="left_attach">1</property>
@@ -283,7 +275,6 @@
                                     <child>
                                       <object class="GtkRadioButton" id="mute_radiobutton">
                                         <property name="label" translatable="yes">Mute Volume</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
                                         <property name="receives_default">False</property>
@@ -300,7 +291,6 @@
                                     <child>
                                       <object class="GtkRadioButton" id="slider_radiobutton">
                                         <property name="label" translatable="yes">Show Slider</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="width_request">77</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
@@ -352,7 +342,6 @@
                                     <child>
                                       <object class="GtkRadioButton" id="mmb_mixer_radiobutton">
                                         <property name="label" translatable="yes">Open Mixer</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
                                         <property name="receives_default">False</property>
@@ -369,7 +358,6 @@
                                     <child>
                                       <object class="GtkRadioButton" id="mmb_mute_radiobutton">
                                         <property name="label" translatable="yes">Mute Volume</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
                                         <property name="receives_default">False</property>
@@ -421,7 +409,6 @@
                                     <child>
                                       <object class="GtkCheckButton" id="use_horizontal_slider">
                                         <property name="label" translatable="yes">Use Horizontal Slider</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
                                         <property name="receives_default">False</property>
@@ -438,7 +425,6 @@
                                     <child>
                                       <object class="GtkCheckButton" id="show_sound_level">
                                         <property name="label" translatable="yes">Show Sound Level</property>
-                                        <property name="use_action_appearance">False</property>
                                         <property name="visible">True</property>
                                         <property name="can_focus">True</property>
                                         <property name="receives_default">False</property>
@@ -472,86 +458,7 @@
                           </packing>
                         </child>
                         <child>
-                          <object class="GtkFrame" id="frame8">
-                            <property name="visible">True</property>
-                            <property name="can_focus">False</property>
-                            <property name="label_xalign">0</property>
-                            <property name="shadow_type">none</property>
-                            <child>
-                              <object class="GtkAlignment" id="alignment8">
-                                <property name="visible">True</property>
-                                <property name="can_focus">False</property>
-                                <property name="left_padding">12</property>
-                                <property name="right_padding">10</property>
-                                <child>
-                                  <object class="GtkTreeView" id="treeview1">
-                                    <property name="visible">True</property>
-                                    <property name="can_focus">True</property>
-                                    <property name="model">hotkey_binding_model</property>
-                                    <property name="headers_visible">False</property>
-                                    <property name="headers_clickable">False</property>
-                                    <property name="fixed_height_mode">True</property>
-                                    <child internal-child="selection">
-                                      <object class="GtkTreeSelection" id="treeview-selection1"/>
-                                    </child>
-                                    <child>
-                                      <object class="GtkTreeViewColumn" id="treeviewcolumn3">
-                                        <property name="sizing">fixed</property>
-                                        <property name="fixed_width">24</property>
-                                        <property name="title" translatable="yes">column</property>
-                                        <child>
-                                          <object class="GtkCellRendererToggle" id="crt_hotkey"/>
-                                          <attributes>
-                                            <attribute name="active">3</attribute>
-                                          </attributes>
-                                        </child>
-                                      </object>
-                                    </child>
-                                    <child>
-                                      <object class="GtkTreeViewColumn" id="treeviewcolumn1">
-                                        <property name="sizing">fixed</property>
-                                        <property name="fixed_width">160</property>
-                                        <property name="title" translatable="yes">column</property>
-                                        <child>
-                                          <object class="GtkCellRendererText" id="cellrenderertext3"/>
-                                          <attributes>
-                                            <attribute name="text">0</attribute>
-                                          </attributes>
-                                        </child>
-                                      </object>
-                                    </child>
-                                    <child>
-                                      <object class="GtkTreeViewColumn" id="treeviewcolumn2">
-                                        <property name="sizing">fixed</property>
-                                        <property name="title" translatable="yes">column2</property>
-                                        <child>
-                                          <object class="GtkCellRendererAccel" id="cra_hotkey">
-                                            <property name="editable">True</property>
-                                          </object>
-                                          <attributes>
-                                            <attribute name="text">1</attribute>
-                                          </attributes>
-                                        </child>
-                                      </object>
-                                    </child>
-                                  </object>
-                                </child>
-                              </object>
-                            </child>
-                            <child type="label">
-                              <object class="GtkLabel" id="label9">
-                                <property name="visible">True</property>
-                                <property name="can_focus">False</property>
-                                <property name="label" translatable="yes">&lt;b&gt;Hotkeys&lt;/b&gt;</property>
-                                <property name="use_markup">True</property>
-                              </object>
-                            </child>
-                          </object>
-                          <packing>
-                            <property name="expand">True</property>
-                            <property name="fill">True</property>
-                            <property name="position">5</property>
-                          </packing>
+                          <placeholder/>
                         </child>
                       </object>
                     </child>
@@ -578,7 +485,6 @@
                 <child>
                   <object class="GtkButton" id="close_button">
                     <property name="label">gtk-close</property>
-                    <property name="use_action_appearance">False</property>
                     <property name="visible">True</property>
                     <property name="can_focus">True</property>
                     <property name="receives_default">True</property>
diff --git a/src/Makefile.am b/src/Makefile.am
index 15060f4..89fcb79 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -15,6 +15,5 @@ volumeicon_SOURCES = \
 	volumeicon.c \
 	config.c \
 	config.h \
-	bind.c \
 	keybinder.h \
 	$(BACKEND)
diff --git a/src/volumeicon.c b/src/volumeicon.c
index 2a1260b..6206168 100644
--- a/src/volumeicon.c
+++ b/src/volumeicon.c
@@ -39,7 +39,6 @@
 #else
 	#include "alsa_backend.h"
 #endif
-#include "keybinder.h"
 #include "config.h"
 
 enum HOTKEY
@@ -249,77 +248,6 @@ static void preferences_mixer_entry_changed(GtkEditable * editable,
 	config_set_helper(gtk_entry_get_text(gui->mixer_entry));
 }
 
-static void preferences_crt_toggled(GtkCellRendererToggle * cell_renderer,
-	gchar * path, gpointer user_data)
-{
-	PreferencesGui * gui = (PreferencesGui*)user_data;
-	GtkTreeIter iter;
-
-	if(gtk_tree_model_get_iter_from_string(GTK_TREE_MODEL(gui->hotkey_store), &iter, path))
-	{
-		gboolean enabled = FALSE;
-		gchar * accel_name = NULL;
-		enum HOTKEY hotkey = 0;
-		gtk_tree_model_get(GTK_TREE_MODEL(gui->hotkey_store), &iter, 1, &accel_name, 2, &hotkey, 3, &enabled, -1);
-		enabled = !enabled;
-
-		if(!enabled)
-			keybinder_unbind(accel_name, hotkey_handle);
-
-		if(enabled && !keybinder_bind(accel_name, hotkey_handle, (void*)hotkey))
-		{
-			g_fprintf(stderr, "Failed to bind %s\n", accel_name);
-		}
-		else
-		{
-			gtk_list_store_set(GTK_LIST_STORE(gui->hotkey_store), &iter, 3, enabled, -1);
-			switch(hotkey)
-			{
-				case UP: config_set_hotkey_up_enabled(enabled); break;
-				case DOWN: config_set_hotkey_down_enabled(enabled); break;
-				case MUTE: config_set_hotkey_mute_enabled(enabled); break;
-			}
-		}
-
-		g_free(accel_name);
-	}
-}
-
-static void preferences_cra_accel_edited(GtkCellRendererAccel * renderer,
-	gchar * path, guint accel_key, GdkModifierType mask, guint hardware_keycode,
-	gpointer user_data)
-{
-	PreferencesGui * gui = (PreferencesGui*)user_data;
-	GtkTreeIter iter;
-	
-	if(gtk_tree_model_get_iter_from_string(GTK_TREE_MODEL(gui->hotkey_store), &iter, path))
-	{
-		enum HOTKEY hotkey = 0;
-		gboolean enabled = FALSE;
-		gchar * old_value = NULL; 
-		gchar * new_value = gtk_accelerator_name(accel_key, mask);
-		gtk_tree_model_get(GTK_TREE_MODEL(gui->hotkey_store), &iter, 1, &old_value, 2, &hotkey, 3, &enabled, -1);
-
-		if(enabled && !keybinder_bind(new_value, hotkey_handle, (void*)hotkey))
-		{
-			g_fprintf(stderr, "Failed to bind %s\n", new_value);
-		}
-		else
-		{
-			gtk_list_store_set(GTK_LIST_STORE(gui->hotkey_store), &iter, 1, new_value, -1);
-			keybinder_unbind(old_value, hotkey_handle);
-			switch(hotkey)
-			{
-				case UP: config_set_hotkey_up(new_value); break;
-				case DOWN: config_set_hotkey_down(new_value); break;
-				case MUTE: config_set_hotkey_mute(new_value); break;
-			}
-		}
-		g_free(new_value);
-		g_free(old_value);
-	}
-}
-
 // Menu handlers
 static void menu_preferences_on_activate(GtkMenuItem * menuitem,
 	gpointer user_data)
@@ -338,7 +266,6 @@ static void menu_preferences_on_activate(GtkMenuItem * menuitem,
 	gui->channel_store = GTK_LIST_STORE(getobj("channel_name_model"));
 	gui->theme_combobox = GTK_COMBO_BOX(getobj("theme_combobox"));
 	gui->theme_store = GTK_LIST_STORE(getobj("theme_name_model"));
-	gui->hotkey_store = GTK_LIST_STORE(getobj("hotkey_binding_model"));
 	gui->close_button = GTK_BUTTON(getobj("close_button"));
 	gui->mute_radiobutton = GTK_RADIO_BUTTON(getobj("mute_radiobutton"));
 	gui->slider_radiobutton = GTK_RADIO_BUTTON(getobj("slider_radiobutton"));
@@ -346,8 +273,6 @@ static void menu_preferences_on_activate(GtkMenuItem * menuitem,
 	gui->mmb_mixer_radiobutton = GTK_RADIO_BUTTON(getobj("mmb_mixer_radiobutton"));
 	gui->use_horizontal_slider_checkbutton = GTK_CHECK_BUTTON(getobj("use_horizontal_slider"));
 	gui->show_sound_level_checkbutton = GTK_CHECK_BUTTON(getobj("show_sound_level"));
-	gui->cra_hotkey = GTK_CELL_RENDERER(getobj("cra_hotkey"));
-	gui->crt_hotkey = GTK_CELL_RENDERER_TOGGLE(getobj("crt_hotkey"));
 	#undef getobj
 
 	// Set the window icon
@@ -411,17 +336,6 @@ static void menu_preferences_on_activate(GtkMenuItem * menuitem,
 	}
 	g_dir_close(themedir);
 
-	// Fill the hotkey binding model
-	gtk_list_store_append(gui->hotkey_store, &tree_iter);
-	gtk_list_store_set(gui->hotkey_store, &tree_iter, 0, "Volume Up", 1,
-		config_get_hotkey_up(), 2, (int)UP, 3, config_get_hotkey_up_enabled(), -1);
-	gtk_list_store_append(gui->hotkey_store, &tree_iter);
-	gtk_list_store_set(gui->hotkey_store, &tree_iter, 0, "Volume Down", 1,
-		config_get_hotkey_down(), 2, (int)DOWN, 3, config_get_hotkey_down_enabled(), -1);
-	gtk_list_store_append(gui->hotkey_store, &tree_iter);
-	gtk_list_store_set(gui->hotkey_store, &tree_iter, 0, "Mute", 1,
-		config_get_hotkey_mute(), 2, (int)MUTE, 3, config_get_hotkey_mute_enabled(), -1);
-
 	// Initialize widgets / connect signals
 	gtk_entry_set_text(gui->mixer_entry, config_get_helper());
 	gtk_adjustment_set_value(gui->volume_adjustment,
@@ -448,10 +362,6 @@ static void menu_preferences_on_activate(GtkMenuItem * menuitem,
 		preferences_use_horizontal_slider_checkbutton_toggled), (gpointer)gui);
 	g_signal_connect(G_OBJECT(gui->show_sound_level_checkbutton), "toggled", G_CALLBACK(
 		preferences_show_sound_level_checkbutton_toggled), (gpointer)gui);
-	g_signal_connect(G_OBJECT(gui->cra_hotkey), "accel-edited", G_CALLBACK(
-		preferences_cra_accel_edited), (gpointer)gui);
-	g_signal_connect(G_OBJECT(gui->crt_hotkey), "toggled", G_CALLBACK(
-		preferences_crt_toggled), (gpointer)gui);
 
 	gtk_widget_show_all(gui->window);
 }
@@ -915,17 +825,6 @@ int main(int argc, char * argv[])
 	status_icon_setup(m_mute);
 	scale_setup();
 
-	keybinder_init();
-	if(config_get_hotkey_up_enabled() &&
-		!keybinder_bind(config_get_hotkey_up(), hotkey_handle, (void*)UP))
-			g_fprintf(stderr, "Failed to bind %s\n", config_get_hotkey_up());
-	if(config_get_hotkey_down_enabled() &&
-		!keybinder_bind(config_get_hotkey_down(), hotkey_handle, (void*)DOWN))
-			g_fprintf(stderr, "Failed to bind %s\n", config_get_hotkey_down());
-	if(config_get_hotkey_mute_enabled() &&
-		!keybinder_bind(config_get_hotkey_mute(), hotkey_handle, (void*)MUTE))
-			g_fprintf(stderr, "Failed to bind %s\n", config_get_hotkey_mute());
-
 	// Main Loop
 	gtk_main();
 
