--- SDL-1.2.14.orig/src/video/fbcon/SDL_fbevents.c	2009-10-13 01:07:15.000000000 +0200
+++ SDL-1.2.14/src/video/fbcon/SDL_fbevents.c	2010-09-16 23:57:26.000000000 +0200

@@ -1001,22 +1001,14 @@
 		TranslateKey(scancode, &keysym);
 		/* Handle Ctrl-Alt-FN for vt switch */
 		switch (keysym.sym) {
-		    case SDLK_F1:
-		    case SDLK_F2:
-		    case SDLK_F3:
-		    case SDLK_F4:
-		    case SDLK_F5:
-		    case SDLK_F6:
-		    case SDLK_F7:
-		    case SDLK_F8:
-		    case SDLK_F9:
-		    case SDLK_F10:
-		    case SDLK_F11:
-		    case SDLK_F12:
+		    case SDLK_1:
+		    case SDLK_2:
+		    case SDLK_3:
+		    case SDLK_4:
 			if ( (SDL_GetModState() & KMOD_CTRL) &&
 			     (SDL_GetModState() & KMOD_ALT) ) {
 				if ( pressed ) {
-					switch_vt(this, (keysym.sym-SDLK_F1)+1);
+					switch_vt(this, (keysym.sym-SDLK_1)+1);
 				}
 				break;
 			}
