diff --git a/src/gmenu2x.cpp b/src/gmenu2x.cpp
index bb5ddee..1c706bc 100644
--- a/src/gmenu2x.cpp
+++ b/src/gmenu2x.cpp
@@ -80,6 +80,13 @@
 #include <sys/types.h>
 #include <dirent.h>
 
+#ifdef PLATFORM_ZAURUS
+#	define UNLOCK_VT
+#	include <sys/ioctl.h>
+#	include <linux/vt.h>
+#	include <linux/kd.h>
+#endif
+
 //for soundcard
 #include <sys/ioctl.h>
 #include <linux/soundcard.h>
@@ -267,16 +274,21 @@ void GMenu2X::init() {
 		f200 = ts.init();
 	}*/
 #else
+#if defined(PLATFORM_ZAURUS)
+	batteryHandle = fopen("/sys/devices/platform/sharpsl-pm/battery_percentage", "r");
+#else
 	batteryHandle = fopen("/sys/class/power_supply/battery/capacity", "r");
-	usbHandle = fopen("/sys/class/power_supply/usb/online", "r");
-	acHandle = fopen("/sys/class/power_supply/ac/online", "r");
-	backlightHandle = fopen(
-#ifdef PLATFORM_NANONOTE
-		"/sys/class/lcd/ili8960-lcd/contrast",
+#endif
+
+#if defined(PLATFORM_NANONOTE)
+	backlightHandle = fopen("/sys/class/lcd/ili8960-lcd/contrast","w+");
+#elif defined(PLATFORM_ZAURUS)
+	backlightHandle = fopen("/sys/class/backlight/corgi_bl/brightness", "w+");
 #else
-		"/sys/class/backlight/pwm-backlight/brightness",
+	backlightHandle = fopen("/sys/class/backlight/pwm-backlight/brightness","w+");
 #endif
-		"w+");
+usbHandle = fopen("/sys/class/power_supply/usb/online", "r");
+acHandle = fopen("/sys/class/power_supply/ac/online", "r");
 #endif
 }
 
