diff --git a/exa/exa_accel.c b/exa/exa_accel.c
index e633d80..735fc49 100644
--- a/exa/exa_accel.c
+++ b/exa/exa_accel.c
@@ -881,7 +881,6 @@ exaImageGlyphBlt (DrawablePtr	pDrawable,
     switch (pDrawable->bitsPerPixel) {
     case 8:	glyph = fbGlyph8; break;
     case 16:    glyph = fbGlyph16; break;
-    case 24:    glyph = fbGlyph24; break;
     case 32:    glyph = fbGlyph32; break;
     }
 
diff --git a/fb/fb.h b/fb/fb.h
index 3742cf0..86f376d 100644
--- a/fb/fb.h
+++ b/fb/fb.h
@@ -69,6 +69,8 @@
 #error "GETLEFTBITS_ALIGNMENT must be 1"
 #endif
 /* whether to bother to include 24bpp support */
+#define FBNO24BIT
+
 #ifndef FBNO24BIT
 #define FB_24BIT
 #endif
diff --git a/hw/kdrive/src/kaa.c b/hw/kdrive/src/kaa.c
index c29e9e9..55bf3cf 100644
--- a/hw/kdrive/src/kaa.c
+++ b/hw/kdrive/src/kaa.c
@@ -791,7 +791,6 @@ kaaImageGlyphBlt (DrawablePtr	pDrawable,
     switch (dstBpp) {
     case 8:	glyph = fbGlyph8; break;
     case 16:    glyph = fbGlyph16; break;
-    case 24:    glyph = fbGlyph24; break;
     case 32:    glyph = fbGlyph32; break;
     }
     
