Index: awesome-3.4.11/screen.c
===================================================================
--- awesome-3.4.11.orig/screen.c	2012-03-12 14:27:38.000000000 +0100
+++ awesome-3.4.11/screen.c	2012-03-12 14:28:29.000000000 +0100
@@ -386,7 +386,7 @@
 static int
 luaA_screen_tags(lua_State *L)
 {
-    screen_t *s = luaL_checkudata(L, 1, "screen");
+    screen_t *s = lua_touserdata(L, 1);
 
     if(lua_gettop(L) == 2)
     {
