diff --git a/hw/kdrive/src/kshadow.c b/hw/kdrive/src/kshadow.c
index 5ca9b2a..8906e01 100644
--- a/hw/kdrive/src/kshadow.c
+++ b/hw/kdrive/src/kshadow.c
@@ -1,5 +1,5 @@
 /*
- * Copyright © 1999 Keith Packard
+ * Copyright ï¿½ 1999 Keith Packard
  *
  * Permission to use, copy, modify, distribute, and sell this software and its
  * documentation for any purpose is hereby granted without fee, provided that
@@ -57,7 +57,6 @@ KdShadowFbFree (KdScreenInfo *screen, int fb)
 {
     if (screen->fb[fb].shadow)
     {
-	xfree (screen->fb[fb].frameBuffer);
 	screen->fb[fb].frameBuffer = 0;
 	screen->fb[fb].shadow = FALSE;
     }
