Index: git/src/ptk/ptk-file-browser.c
===================================================================
--- git.orig/src/ptk/ptk-file-browser.c	2012-04-01 02:11:16.000000000 +0200
+++ git/src/ptk/ptk-file-browser.c	2012-04-01 02:13:17.000000000 +0200
@@ -1330,6 +1330,7 @@
                                                                 FALSE, FALSE, 0 );
     gtk_label_set_selectable( file_browser->status_label, TRUE ); // required for button event
-    gtk_widget_set_can_focus( file_browser->status_label, FALSE );
+    GTK_WIDGET_UNSET_FLAGS (file_browser->status_label, GTK_CAN_FOCUS);
     g_signal_connect( G_OBJECT( file_browser->status_label ), "button-press-event",
                       G_CALLBACK( on_status_bar_button_press ), file_browser );
     g_signal_connect( G_OBJECT( file_browser->status_label ), "populate-popup",
Index: git/src/settings.c
===================================================================
--- git.orig/src/settings.c	2012-04-01 02:11:20.000000000 +0200
+++ git/src/settings.c	2012-04-01 02:13:34.000000000 +0200
@@ -7041,7 +7041,7 @@
         gtk_button_set_image( btn_help, xset_get_image( "GTK_STOCK_HELP",
                                                             GTK_ICON_SIZE_BUTTON ) );
         gtk_button_set_focus_on_click( btn_help, FALSE );
-        gtk_widget_set_can_focus( btn_help, FALSE );
+        GTK_WIDGET_UNSET_FLAGS (btn_help, GTK_CAN_FOCUS);
     }
 
     gtk_widget_show_all( dlg );
